File: src/lib/tools/tavilySearch.ts
// Tool: Tavily Web Search
// Wraps the @tavily/core package for use inside LangGraph nodes.

import { tavily } from "@tavily/core";

// ─── Types ───────────────────────────────────────────────────────────────────

export interface TavilyResult {
  title: string;
  url: string;
  content: string;
  score: number;
}

// ─── Client Factory ──────────────────────────────────────────────────────────

const getClient = () => {
  const apiKey = process.env.TAVILY_API_KEY;
  if (!apiKey) throw new Error("TAVILY_API_KEY is not set in environment variables.");
  return tavily({ apiKey });
};

// ─── runTavilySearch ─────────────────────────────────────────────────────────
// Runs a single Tavily search and returns structured results.
// On failure, logs the error and returns an empty array (never throws).

export const runTavilySearch = async (
  query: string,
  maxResults: number = 5
): Promise<TavilyResult[]> => {
  try {
    const client = getClient();

    const response = await client.search(query, {
      searchDepth: "advanced",
      includeAnswer: true,
      maxResults,
    });

    const results: TavilyResult[] = (response.results ?? []).map((r) => ({
      title: r.title ?? "",
      url: r.url ?? "",
      content: r.content ?? "",
      score: r.score ?? 0,
    }));

    return results;
  } catch (error) {
    console.error(`[TavilySearch] Failed for query "${query}":`, error);
    return [];
  }
};

// ─── runMultipleSearches ─────────────────────────────────────────────────────
// Runs multiple queries in parallel, merges all results into a flat array,
// and deduplicates by URL before returning.

export const runMultipleSearches = async (
  queries: string[],
  maxResultsPerQuery: number = 5
): Promise<TavilyResult[]> => {
  const settled = await Promise.allSettled(
    queries.map((q) => runTavilySearch(q, maxResultsPerQuery))
  );

  const allResults: TavilyResult[] = [];
  for (const result of settled) {
    if (result.status === "fulfilled") {
      allResults.push(...result.value);
    }
    // rejected means runTavilySearch already handled the error and returned []
    // so this branch is unreachable in practice, but we're defensive here
  }

  // Deduplicate by URL, keeping the highest-scoring entry for each URL
  const seen = new Map<string, TavilyResult>();
  for (const item of allResults) {
    const existing = seen.get(item.url);
    if (!existing || item.score > existing.score) {
      seen.set(item.url, item);
    }
  }

  // Sort descending by score before returning
  return Array.from(seen.values()).sort((a, b) => b.score - a.score);
};

File: backend/tools/tavily_search.py
"""
tavily_search.py — Tavily Web Search tool
Port of src/lib/tools/tavilySearch.ts
"""

import asyncio
import logging
import os
from typing import Optional

logger = logging.getLogger(__name__)


def _get_client():
    """Build a synchronous Tavily client (used inside thread executor)."""
    from tavily import TavilyClient

    api_key = os.environ.get("TAVILY_API_KEY")
    if not api_key:
        raise ValueError("TAVILY_API_KEY is not set in environment variables.")
    return TavilyClient(api_key=api_key)


async def run_tavily_search(query: str, max_results: int = 5) -> list[dict]:
    """
    Run a single Tavily search asynchronously and return structured results.
    On failure, logs the error and returns an empty list (never raises).
    """
    try:
        client = _get_client()
        response = await asyncio.to_thread(
            client.search,
            query,
            search_depth="advanced",
            include_answer=True,
            max_results=max_results,
        )
        results = response.get("results") or []
        return [
            {
                "title": r.get("title", ""),
                "url": r.get("url", ""),
                "content": r.get("content", ""),
                "score": r.get("score", 0),
            }
            for r in results
        ]
    except Exception as exc:
        logger.error('[TavilySearch] Failed for query "%s": %s', query, exc)
        return []


async def run_multiple_searches(
    queries: list[str], max_results_per_query: int = 5
) -> list[dict]:
    """
    Run multiple queries in parallel, merge all results into a flat list,
    and deduplicate by URL (keeping the highest-scoring entry per URL).
    """
    tasks = [run_tavily_search(q, max_results_per_query) for q in queries]
    settled = await asyncio.gather(*tasks, return_exceptions=True)

    all_results: list[dict] = []
    for result in settled:
        if isinstance(result, list):
            all_results.extend(result)
        # exceptions already handled inside run_tavily_search

    # Deduplicate by URL, keeping highest score
    seen: dict[str, dict] = {}
    for item in all_results:
        url = item["url"]
        if url not in seen or item["score"] > seen[url]["score"]:
            seen[url] = item

    # Sort descending by score
    return sorted(seen.values(), key=lambda x: x["score"], reverse=True)
